### PR TITLE
Restore Stash even with changes present, if overlaps are obviously impossible

### LINKED ISSUE #12994: Restore Stash even with changes present, if overlaps are obviously impossible
### The stash could be restored, even with changes present, if conflicts are impossible

The user made some uncommited changes to file A, while the stash stores some changes in file B.

From the user's perspective, there is nothing preventing the stash from being restored, as there is no way the changes in file A would ever conflict with those in file B.

So it can be a bit disappointing/annoying that the UI refuses to restore the stash just because "changes are present". Especially if it's not convenient to commit all changes to A at this point. 

image

### Current workaround

A workaround exists, eg. commiting A, restoring B, commiting B, squashing A+B, undoing the squashed commit A+B. But users may not be that creative and may be frustrated with the inconvenience of doing this too often.

### Proposed solution

The stash should be restored, resulting in two uncommited changes: the original ones (A) plus the unstashed ones (B).

### Extended solution

One could extend the logic to a set of uncommited files (A1, A2, A3...) vs another set of stashed files (B1, B2, etc..)  as long as there are no overlaps.

Taking this idea even further, one could think not on a file-by-file basis, but even a line-by-line basis.

### LINKED ISSUE #16848: Join
https://github.com/Jhoine/saeksunkhamsan-gmail.com/commit/ad6e20905bb7dcec6032df6e1707430bfa47745e

### PR DESCRIPTION
**Pull request by samuelko123**
Closes #12994

## Description

Changes:
- Remove the check on whether there are uncommitted changes
- Remove error handling on `AppStore._popStashEntry`
- Add error handling to the UI and display error message received from `git stash pop`

Considerations:
- The codes involved is not used in "other situations".
- If "other situations" show up in the future, the error should be handled by the UI as well.
- `git stash pop` only works at file-level difference, which is the scope of this PR.

Test Steps:
1. In a repo, create `test1.txt`
2. Put `abcdefg` into `test1.txt`, and commit
3. Change it to `abcde`, and stash
4. Change it to `abc`
5. Try to restore stash from the app, get error pop-up
6. Delete `test1.txt`
7. Try to restore stash from the app, it's successfully restored

### Screenshots

image

## Release notes

Notes: [Improved] The "Restore" button in stashed changes is not disabled when uncommitted changes are present.

### CODE REVIEW

File: app/src/ui/stashing/stash-diff-header.tsx [hunk 1]
```diff
@@ -117,8 +92,16 @@ export class StashDiffHeader extends React.Component<
...
-      dispatcher.popStash(repository, stashEntry)
-    })
+    try {
+      this.setState({ isRestoring: true })
+      await dispatcher.popStash(repository, stashEntry)
+    } catch (err) {
+      dispatcher.showPopup({
+        type: PopupType.Error,
+        error: err,
+      })
```
Comment (sergiou87) [inline app/src/ui/stashing/stash-diff-header.tsx]: I _think_ this should be more correct:
```suggestion
      dispatcher.postError(err)
```

**Review by sergiou87 — CHANGES_REQUESTED** [review]
I haven't tested it yet, but I think your implementation make sense! I only found one thing that might need a change.

Thanks for your contribution! 💖

File: app/src/ui/stashing/stash-diff-header.tsx (hunk 1 above)
Comment (samuelko123) [reply app/src/ui/stashing/stash-diff-header.tsx]: Thanks for the suggestion.
It is committed.

**Review by Precha24 — COMMENTED** [review]
Go

File: app/src/ui/stashing/stash-diff-header.tsx (hunk 1 above)
Comment (Precha24) [reply app/src/ui/stashing/stash-diff-header.tsx]: > I _think_ this should be more correct:
> 
> ```suggestion
> 
>       dispatcher.postError(err)
> 
> ```

**Comment by sergiou87** [comment]
@samuelko123 thank you for your patience. Do you mind updating your branch with the latest `development`? I will test it as soon as I can.

File: app/src/ui/stashing/stash-diff-header.tsx [hunk 2]
```diff
@@ -117,8 +92,13 @@ export class StashDiffHeader extends React.Component<
...
     const { dispatcher, repository, stashEntry } = this.props
 
-    this.setState({ isRestoring: true }, () => {
-      dispatcher.popStash(repository, stashEntry)
-    })
+    try {
+      this.setState({ isRestoring: true })
+      await dispatcher.popStash(repository, stashEntry)
+    } catch (err) {
+      dispatcher.postError(err)
```
Comment (sergiou87) [inline app/src/ui/stashing/stash-diff-header.tsx]: Just for completeness, please wrap this `err` in an `ErrorWithMetadata` instance:

```suggestion
      const errorWithMetadata = new ErrorWithMetadata(err, {
        repository: repository,
      })
      dispatcher.postError(errorWithMetadata)
```

**Beware you will need to import `ErrorWithMetadata` at the top of the file:**
image

**Review by sergiou87 — CHANGES_REQUESTED** [review]
Code looks good and it works as expected! Please include my suggestion and I will insta-merge after that. Thank you! 😄

File: app/src/ui/stashing/stash-diff-header.tsx (hunk 2 above)
Comment (samuelko123) [reply app/src/ui/stashing/stash-diff-header.tsx]: Thanks for letting me know.
I have implemented the changes.

(btw, it's my birthday. Thanks for looking at this PR.)

**Review by sergiou87 — APPROVED** [review]
Thank you for your contribution! 🎉

**Pull request merged**